README.md
# Helix JSON Filter

> Service used to filter JSON data from a helix3 content bus

## Status
[![codecov](https://img.shields.io/codecov/c/github/adobe/helix-json-filter.svg)](https://codecov.io/gh/adobe/helix-json-filter)
[![CircleCI](https://img.shields.io/circleci/project/github/adobe/helix-json-filter.svg)](https://circleci.com/gh/adobe/helix-json-filter)
[![GitHub license](https://img.shields.io/github/license/adobe/helix-json-filter.svg)](https://github.com/adobe/helix-json-filter/blob/main/LICENSE.txt)
[![GitHub issues](https://img.shields.io/github/issues/adobe/helix-json-filter.svg)](https://github.com/adobe/helix-json-filter/issues)
[![LGTM Code Quality Grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/adobe/helix-json-filter.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/adobe/helix-json-filter)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)

## Installation

## Usage

```bash
curl https://helix-pages.anywhere.run/helix-services/json-filter@v1
```

For more, see the [API documentation](docs/API.md).

## Development

Please note that helix3 functions require authentication. in order to deploy manually and run the deploy tests, 
add the authorization to the test headers to your `.env` file. eg:

```
HLX_TEST_HEADERS={"authorization": "token xxxxxxxxxxyyyyyyyyyzzzzzzz"}
```

### Deploying Helix JSON Filter

All commits to main that pass the testing will be deployed automatically. All commits to branches that will pass the testing will get commited as `/helix-services/json-filter@ci<num>` and tagged with the CI build number.


CHANGELOG.md
# [3.0.0](https://github.com/adobe/helix-json-filter/compare/v2.0.2...v3.0.0) (2021-08-10)


### Features

* use new path format with contentBusPartition as path segment ([#47](https://github.com/adobe/helix-json-filter/issues/47)) ([ea44d86](https://github.com/adobe/helix-json-filter/commit/ea44d86eab8bfd30322b9e4891981f0805dfae60))


### BREAKING CHANGES

* in order to have a similar path format as the
  pipeline-serivice, the json-filter requires now also the
  contentBusPartition as path segment, eg: /live/path

## [2.0.2](https://github.com/adobe/helix-json-filter/compare/v2.0.1...v2.0.2) (2021-08-07)


### Bug Fixes

* **deps:** update external fixes ([#45](https://github.com/adobe/helix-json-filter/issues/45)) ([a2667b5](https://github.com/adobe/helix-json-filter/commit/a2667b57a158547bc388aa2840a51cddc63e5dd6))

## [2.0.1](https://github.com/adobe/helix-json-filter/compare/v2.0.0...v2.0.1) (2021-08-07)


### Bug Fixes

* unly use aws monitor target ([#44](https://github.com/adobe/helix-json-filter/issues/44)) ([f5ece76](https://github.com/adobe/helix-json-filter/commit/f5ece76208d9d82b28b2dd797c326c799833de99))

# [2.0.0](https://github.com/adobe/helix-json-filter/compare/v1.1.16...v2.0.0) (2021-08-04)


### Bug Fixes

* use single bucket for content bus ([#37](https://github.com/adobe/helix-json-filter/issues/37)) ([a22fe07](https://github.com/adobe/helix-json-filter/commit/a22fe0708a767c4e5d85d918fb9d06718e45b2b0)), closes [#36](https://github.com/adobe/helix-json-filter/issues/36)


### BREAKING CHANGES

* content-bus id no longer addresses the bucket 
  but is a prefix for the key in the helix-content-bus bucket

## [1.1.16](https://github.com/adobe/helix-json-filter/compare/v1.1.15...v1.1.16) (2021-08-01)


### Bug Fixes

* **deps:** update external fixes ([#43](https://github.com/adobe/helix-json-filter/issues/43)) ([8abb299](https://github.com/adobe/helix-json-filter/commit/8abb29948587a0916b819f51a15fc3af0ce6645b))

## [1.1.15](https://github.com/adobe/helix-json-filter/compare/v1.1.14...v1.1.15) (2021-07-30)


### Bug Fixes

* **deps:** update adobe fixes ([#41](https://github.com/adobe/helix-json-filter/issues/41)) ([af0a624](https://github.com/adobe/helix-json-filter/commit/af0a6240455644a874e04c3da8ad679885816348))

## [1.1.14](https://github.com/adobe/helix-json-filter/compare/v1.1.13...v1.1.14) (2021-07-30)


### Bug Fixes

* **deps:** update dependency @adobe/helix-fetch to v2.4.0 ([#40](https://github.com/adobe/helix-json-filter/issues/40)) ([a6b97eb](https://github.com/adobe/helix-json-filter/commit/a6b97eb013731b1c3a38e3d0a6476626aa2cdacc))

## [1.1.13](https://github.com/adobe/helix-json-filter/compare/v1.1.12...v1.1.13) (2021-07-28)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.2.0 ([#38](https://github.com/adobe/helix-json-filter/issues/38)) ([84139f5](https://github.com/adobe/helix-json-filter/commit/84139f531a7109cbd4de6119a4cff947f5ab76cb))

## [1.1.12](https://github.com/adobe/helix-json-filter/compare/v1.1.11...v1.1.12) (2021-07-24)


### Bug Fixes

* **deps:** update external fixes ([#35](https://github.com/adobe/helix-json-filter/issues/35)) ([6395b5d](https://github.com/adobe/helix-json-filter/commit/6395b5d7cbbaeb8b1ac9c54001a2d85a33d98bfc))

## [1.1.11](https://github.com/adobe/helix-json-filter/compare/v1.1.10...v1.1.11) (2021-07-18)


### Bug Fixes

* **deps:** update external fixes ([#32](https://github.com/adobe/helix-json-filter/issues/32)) ([d702afb](https://github.com/adobe/helix-json-filter/commit/d702afb5cacdacce238ee9ca8796601de60915c1))

## [1.1.10](https://github.com/adobe/helix-json-filter/compare/v1.1.9...v1.1.10) (2021-07-14)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.5 ([#31](https://github.com/adobe/helix-json-filter/issues/31)) ([95f4951](https://github.com/adobe/helix-json-filter/commit/95f4951dad464ec07113bbaab384ebed10ece44e))

## [1.1.9](https://github.com/adobe/helix-json-filter/compare/v1.1.8...v1.1.9) (2021-07-11)


### Bug Fixes

* **deps:** update dependency @aws-sdk/client-s3 to v3.21.0 ([#30](https://github.com/adobe/helix-json-filter/issues/30)) ([20fb278](https://github.com/adobe/helix-json-filter/commit/20fb278881d26ce55e4cf1a59454df63bd5e43d7))

## [1.1.8](https://github.com/adobe/helix-json-filter/compare/v1.1.7...v1.1.8) (2021-07-07)


### Bug Fixes

* **deps:** update dependency @adobe/helix-universal to v1.6.4 ([#29](https://github.com/adobe/helix-json-filter/issues/29)) ([86f3ab2](https://github.com/adobe/helix-json-filter/commit/86f3ab22fdd9a6be8221052e207bbd755b7a3324))

## [1.1.7](https://github.com/adobe/helix-json-filter/compare/v1.1.6...v1.1.7) (2021-07-05)


### Bug Fixes

* **deps:** update external fixes ([#27](https://github.com/adobe/helix-json-filter/issues/27)) ([6266da3](https://github.com/adobe/helix-json-filter/commit/6266da306f1cf785821837011f5a40bd3aa9fee9))

## [1.1.6](https://github.com/adobe/helix-json-filter/compare/v1.1.5...v1.1.6) (2021-06-29)


### Bug Fixes

* **deps:** update dependency @aws-sdk/client-s3 to v3.19.0 ([8df1bb1](https://github.com/adobe/helix-json-filter/commit/8df1bb17104e118e4cf79e2946765125771bfa6a))

## [1.1.5](https://github.com/adobe/helix-json-filter/compare/v1.1.4...v1.1.5) (2021-06-13)


### Bug Fixes

* **deps:** update dependency @adobe/helix-status to v9.1.8 ([#23](https://github.com/adobe/helix-json-filter/issues/23)) ([ea6ea07](https://github.com/adobe/helix-json-filter/commit/ea6ea07000622565a51a28ae703fcd91fc34a4ae))

## [1.1.4](https://github.com/adobe/helix-json-filter/compare/v1.1.3...v1.1.4) (2021-06-13)


### Bug Fixes

* **deps:** update dependency @adobe/helix-fetch to v2.3.0 ([#21](https://github.com/adobe/helix-json-filter/issues/21)) ([9caf82b](https://github.com/adobe/helix-json-filter/commit/9caf82b88e1e5fce3a54bbed6aee4e21488c801d))

## [1.1.3](https://github.com/adobe/helix-json-filter/compare/v1.1.2...v1.1.3) (2021-06-09)


### Bug Fixes


test/index.test.js
/*
 * Copyright 2021 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
/* eslint-env mocha */
const assert = require('assert');
const { Request, Response } = require('@adobe/helix-universal');
const proxyquire = require('proxyquire');

const { main } = require('../src/index.js');

const TEST_DATA = require('./fixtures/test-data.json');
const TEST_DATA_INVALID = require('./fixtures/test-data-invalid.json');

const TEST_SINGLE_SHEET = {
  offset: 0,
  limit: TEST_DATA.length,
  total: TEST_DATA.length,
  data: TEST_DATA,
};

describe('Index Tests', () => {
  it('sends 400 for missing contentBusId', async () => {
    const result = await main(new Request('https://json-filter.com/'), { log: console, pathInfo: {} });
    assert.strictEqual(result.status, 400);
  });

  it('sends 400 for missing suffix', async () => {
    const result = await main(new Request('https://json-filter.com/?contentBusId=foobar'), {
      log: console,
      pathInfo: {
      },
    });
    assert.strictEqual(result.status, 400);
  });

  it('sends 400 for non json  suffix', async () => {
    const result = await main(new Request('https://json-filter.com/?contentBusId=foobar'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.md',
      },
    });
    assert.strictEqual(result.status, 400);
  });

  it('sends 404 for missing query params', async () => {
    const result = await main(new Request('https://json-filter.com/?contentBusId=foobar'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.json',
      },
    });
    assert.strictEqual(result.status, 400);
  });

  it('fetches correct content', async () => {
    const { main: proxyMain } = proxyquire('../src/index.js', {
      './fetch-s3.js': async (bucketId, key) => {
        assert.strictEqual(bucketId, 'helix-content-bus');
        assert.strictEqual(key, 'foobar/preview/index.json');
        return new Response(JSON.stringify(TEST_SINGLE_SHEET), {
          status: 200,
          headers: {
            'content-type': 'application/json',
          },
        });
      },
    });

    const resp = await proxyMain(new Request('https://json-filter.com/?contentBusId=foobar&limit=10&offset=5'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.json',
      },
    });
    assert.strictEqual(resp.status, 200);
    assert.deepStrictEqual(await resp.json(), {
      ':type': 'sheet',
      offset: 5,
      limit: 10,
      total: TEST_DATA.length,
      data: TEST_DATA.slice(5, 15),
    });
    assert.deepStrictEqual(resp.headers.plain(), {
      'content-type': 'application/json',
    });
  });

  it('handles error from content', async () => {
    const { main: proxyMain } = proxyquire('../src/index.js', {
      './fetch-s3.js': async () => new Response('', { status: 404 }),
    });

    const resp = await proxyMain(new Request('https://json-filter.com/?contentBusId=foobar&limit=10&offset=5'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.json',
      },
    });
    assert.strictEqual(resp.status, 404);
  });

  it('handles error from filter', async () => {
    const { main: proxyMain } = proxyquire('../src/index.js', {
      './fetch-s3.js': async () => new Response(JSON.stringify(TEST_SINGLE_SHEET), {
        status: 200,
        headers: {
          'content-type': 'application/json',
        },
      }),
      './json-filter.js': () => () => new Response('', { status: 404 }),
    });

    const resp = await proxyMain(new Request('https://json-filter.com/?contentBusId=foobar&limit=10&offset=5'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.json',
      },
    });
    assert.strictEqual(resp.status, 404);
  });

  it('creates correct filter with no offset', async () => {
    const { main: proxyMain } = proxyquire('../src/index.js', {
      './fetch-s3.js': async () => new Response(JSON.stringify(TEST_SINGLE_SHEET), {
        status: 200,
        headers: {
          'content-type': 'application/json',
        },
      }),
      './json-filter.js': (query) => {
        assert.strictEqual(query.limit, 10);
        assert.strictEqual(query.offset, undefined);
        assert.strictEqual(query.sheet, undefined);
        return () => new Response('', { status: 200 });
      },
    });

    const resp = await proxyMain(new Request('https://json-filter.com/?contentBusId=foobar&limit=10'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.json',
      },
    });
    assert.strictEqual(resp.status, 200);
  });

  it('creates correct filter with no limit', async () => {
    const { main: proxyMain } = proxyquire('../src/index.js', {
      './fetch-s3.js': async () => new Response(JSON.stringify(TEST_SINGLE_SHEET), {
        status: 200,
        headers: {
          'content-type': 'application/json',
        },
      }),
      './json-filter.js': (query) => {
        assert.strictEqual(query.limit, undefined);
        assert.strictEqual(query.offset, 10);
        assert.strictEqual(query.sheet, undefined);
        return () => new Response('', { status: 200 });
      },
    });

    const resp = await proxyMain(new Request('https://json-filter.com/?contentBusId=foobar&offset=10'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.json',
      },
    });
    assert.strictEqual(resp.status, 200);
  });

  it('creates correct filter with multiple sheets', async () => {
    const { main: proxyMain } = proxyquire('../src/index.js', {
      './fetch-s3.js': async () => new Response(JSON.stringify(TEST_SINGLE_SHEET), {
        status: 200,
        headers: {
          'content-type': 'application/json',
        },
      }),
      './json-filter.js': (query) => {
        assert.strictEqual(query.limit, undefined);
        assert.strictEqual(query.offset, undefined);
        assert.deepStrictEqual(query.sheet, ['sheet1', 'sheet2']);
        return () => new Response('', { status: 200 });
      },
    });

    const resp = await proxyMain(new Request('https://json-filter.com/?contentBusId=foobar&sheet=sheet1&sheet=sheet2'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.json',
      },
    });
    assert.strictEqual(resp.status, 200);
  });

  it('creates correct filter with single sheet', async () => {
    const { main: proxyMain } = proxyquire('../src/index.js', {
      './fetch-s3.js': async () => new Response(JSON.stringify(TEST_SINGLE_SHEET), {
        status: 200,
        headers: {
          'content-type': 'application/json',
        },
      }),
      './json-filter.js': (query) => {
        assert.strictEqual(query.limit, undefined);
        assert.strictEqual(query.offset, undefined);
        assert.deepStrictEqual(query.sheet, ['sheet1']);
        return () => new Response('', { status: 200 });
      },
    });

    const resp = await proxyMain(new Request('https://json-filter.com/?contentBusId=foobar&sheet=sheet1'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.json',
      },
    });
    assert.strictEqual(resp.status, 200);
  });

  it('handles multi-sheet with missing :names property', async () => {
    const { main: proxyMain } = proxyquire('../src/index.js', {
      './fetch-s3.js': async () => new Response(JSON.stringify(TEST_DATA_INVALID), {
        status: 200,
        headers: {
          'content-type': 'application/json',
        },
      }),
    });

    const resp = await proxyMain(new Request('https://json-filter.com/?contentBusId=foobar&sheet=countries&offset=2&limit=1'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.json',
      },
    });
    assert.strictEqual(resp.status, 502);
    assert.strictEqual(resp.headers.get('x-error'), 'multisheet data invalid. missing ":names" property.');
  });

  it('handles corrupt json', async () => {
    const { main: proxyMain } = proxyquire('../src/index.js', {
      './fetch-s3.js': async () => new Response('this is not json!', {
        status: 200,
        headers: {
          'content-type': 'application/json',
        },
      }),
    });

    const resp = await proxyMain(new Request('https://json-filter.com/?contentBusId=foobar&sheet=countries&offset=2&limit=1'), {
      log: console,
      pathInfo: {
        suffix: '/preview/index.json',
      },
    });
    assert.strictEqual(resp.status, 502);
    assert.strictEqual(resp.headers.get('x-error'), 'failed to parse json: Unexpected token h in JSON at position 1');
  });
});


test/post-deploy.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */
/* eslint-disable no-unused-expressions */

const querystring = require('querystring');
const chai = require('chai');
const chaiHttp = require('chai-http');
const { createTargets } = require('./post-deploy-utils.js');

chai.use(chaiHttp);
const { expect } = chai;

const SINGLE_SHEET_PATH = '/data-embed-unit-tests/example-data.json';

const MULTI_SHEET_PATH = '/data-embed-unit-tests/multisheet.json';

const DEFAULT_PARAMS = {
  // https://adobe.sharepoint.com/sites/cg-helix/Shared%20Documents/Forms/AllItems.aspx
  contentBusId: 'd4b1d4ea6d84b0229bce7cf6806b0bb3470489ab8205a13f75cfe518fa7',
};

createTargets().forEach((target) => {
  describe(`Post-Deploy Tests (${target.title()})`, () => {
    it('filters a single sheet', async () => {
      const path = `${target.urlPath()}/live${SINGLE_SHEET_PATH}?${querystring.encode({
        ...DEFAULT_PARAMS,
        limit: 4,
      })}`;

      const req = chai
        .request(target.host())
        .get(path);

      // todo: move to utils ?
      if (process.env.HLX_TEST_HEADERS) {
        const headers = JSON.parse(process.env.HLX_TEST_HEADERS);
        Object.entries(headers).forEach(([key, value]) => req.set(key, value));
      }
      await req
        .then((response) => {
          expect(response).to.have.status(200);
          expect(response.body).to.deep.eql({
            ':type': 'sheet',
            data: [
              { Code: 'JP', Country: 'Japan', Number: 3 },
              { Code: 'DE', Country: 'Germany', Number: 5 },
              { Code: 'US', Country: 'USA', Number: 7 },
              { Code: 'CH', Country: 'Switzerland', Number: 27 },
            ],
            limit: 4,
            offset: 0,
            total: 6,
          });
        }).catch((e) => {
          throw e;
        });
    }).timeout(50000);

    it('filters a multi sheet', async () => {
      const path = `${target.urlPath()}/live${MULTI_SHEET_PATH}?${querystring.encode({
        ...DEFAULT_PARAMS,
        limit: 1,
      })}`;

      const req = chai
        .request(target.host())
        .get(path);

      if (process.env.HLX_TEST_HEADERS) {
        const headers = JSON.parse(process.env.HLX_TEST_HEADERS);
        Object.entries(headers).forEach(([key, value]) => req.set(key, value));
      }
      await req
        .then((response) => {
          expect(response).to.have.status(200);
          expect(response.body).to.deep.eql({
            ':names': [
              '日本',
              'tables',
              'foo',
            ],
            ':type': 'multi-sheet',
            ':version': 3,
            foo: {
              data: [{ Code: 'JP', Country: 'Japan', Number: 3 }],
              limit: 1,
              offset: 0,
              total: 6,
            },
            tables: {
              data: [{
                '': '', A: 112, B: 224, C: 135,
              }],
              limit: 1,
              offset: 0,
              total: 10,
            },
            日本: {
              data: [{ Code: 'JP', Country: 'Japan', Number: 3 }],
              limit: 1,
              offset: 0,
              total: 1,
            },
          });
        }).catch((e) => {
          throw e;
        });
    }).timeout(50000);

    it('selects single sheet from a multi sheet', async () => {
      const path = `${target.urlPath()}/live${MULTI_SHEET_PATH}?${querystring.encode({
        ...DEFAULT_PARAMS,
        limit: 1,
        sheet: '日本',
      })}`;

      const req = chai
        .request(target.host())
        .get(path);

      if (process.env.HLX_TEST_HEADERS) {
        const headers = JSON.parse(process.env.HLX_TEST_HEADERS);
        Object.entries(headers).forEach(([key, value]) => req.set(key, value));
      }
      await req
        .then((response) => {
          expect(response).to.have.status(200);
          expect(response.body).to.deep.eql({
            ':type': 'sheet',
            data: [
              { Code: 'JP', Country: 'Japan', Number: 3 },
            ],
            limit: 1,
            offset: 0,
            total: 1,
          });
        }).catch((e) => {
          throw e;
        });
    }).timeout(50000);

    it('selects wrong sheet from a multi sheet', async () => {
      const path = `${target.urlPath()}/live${MULTI_SHEET_PATH}?${querystring.encode({
        ...DEFAULT_PARAMS,
        limit: 1,
        sheet: 'does-not-exist',
      })}`;

      const req = chai
        .request(target.host())
        .get(path);

      if (process.env.HLX_TEST_HEADERS) {
        const headers = JSON.parse(process.env.HLX_TEST_HEADERS);
        Object.entries(headers).forEach(([key, value]) => req.set(key, value));
      }
      await req
        .then((response) => {
          expect(response).to.have.status(404);
          expect(response.headers['x-error']).to.eql('filtered result does not contain selected sheet(s): does-not-exist');
        }).catch((e) => {
          throw e;
        });
    }).timeout(50000);

    /**
     * Since we don't know the sheet name of the original single sheet, using a sheet selector
     * is irrelevant.
     * alternatively, it could always respond with 404 if a sheet selector is specified.
     */
    it('selects sheet from a single sheet', async () => {
      const path = `${target.urlPath()}/live${SINGLE_SHEET_PATH}?${querystring.encode({
        ...DEFAULT_PARAMS,
        limit: 1,
        sheet: 'does-not-exist',
      })}`;

      const req = chai
        .request(target.host())
        .get(path);

      // todo: move to utils ?
      if (process.env.HLX_TEST_HEADERS) {
        const headers = JSON.parse(process.env.HLX_TEST_HEADERS);
        Object.entries(headers).forEach(([key, value]) => req.set(key, value));
      }
      await req
        .then((response) => {
          expect(response).to.have.status(200);
          expect(response.body).to.deep.eql({
            ':type': 'sheet',
            data: [
              { Code: 'JP', Country: 'Japan', Number: 3 },
            ],
            limit: 1,
            offset: 0,
            total: 6,
          });
        }).catch((e) => {
          throw e;
        });
    }).timeout(50000);
  });
});


test/fetch-s3.test.js
/*
 * Copyright 2021 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */

/* eslint-env mocha */

const assert = require('assert');
const nock = require('nock');
const { promisify } = require('util');
const zlib = require('zlib');

const fetchS3 = require('../src/fetch-s3.js');

const gzip = promisify(zlib.gzip);

describe('Fetch S3 test', () => {
  beforeEach(() => {
    process.env.AWS_REGION = 'fake';
    process.env.AWS_ACCESS_KEY_ID = 'fake-key-id';
    process.env.AWS_SECRET_ACCESS_KEY = 'fake-secret';
  });

  it('needs bucket id', async () => {
    await assert.rejects(fetchS3(), new Error('Unknown bucketId, cannot fetch content'));
  });

  it('needs key', async () => {
    await assert.rejects(fetchS3('test-bucket'), new Error('Unknown key, cannot fetch content'));
  });

  it('can get content', async () => {
    const scope = nock('https://test-bucket.s3.fake.amazonaws.com')
      .get('/live/index.json?x-id=GetObject')
      .reply(200, '{ "welcome": 42 }', {
        'content-type': 'application/json',
        'last-modified': 'Wed, 12 Oct 2009 17:50:00 GMT',
      });

    const resp = await fetchS3('test-bucket', 'live/index.json', console);
    await scope.done();

    assert.strictEqual(resp.status, 200);
    assert.deepStrictEqual(await resp.json(), { welcome: 42 });
  });

  it('can get content w/o type', async () => {
    const scope = nock('https://test-bucket.s3.fake.amazonaws.com')
      .get('/live/index.json?x-id=GetObject')
      .reply(200, '{ "welcome": 42 }');

    const resp = await fetchS3('test-bucket', 'live/index.json', console);
    await scope.done();

    assert.strictEqual(resp.status, 200);
    assert.deepStrictEqual(await resp.json(), { welcome: 42 });
  });

  it('can get zipped content', async () => {
    const data = await gzip(Buffer.from('{ "welcome": 42 }', 'utf-8'));

    const scope = nock('https://test-bucket.s3.fake.amazonaws.com')
      .get('/live/index.json?x-id=GetObject')
      .reply(200, data, {
        'content-type': 'application/json',
        'content-encoding': 'gzip',
        'last-modified': 'Wed, 12 Oct 2009 17:50:00 GMT',
      });

    const resp = await fetchS3('test-bucket', 'live/index.json', console);
    await scope.done();

    assert.strictEqual(resp.status, 200);
    assert.deepStrictEqual(await resp.json(), { welcome: 42 });
  });

  it('handles 404 from s3', async () => {
    const scope = nock('https://test-bucket.s3.fake.amazonaws.com')
      .get('/live/index.json?x-id=GetObject')
      .reply(404);

    const resp = await fetchS3('test-bucket', 'live/index.json', console);
    await scope.done();

    assert.strictEqual(resp.status, 404);
  });

  it('handles AccessDenied from s3', async () => {
    const scope = nock('https://test-bucket.s3.fake.amazonaws.com')
      .get('/live/index.json?x-id=GetObject')
      .reply(403, '<?xml version="1.0" encoding="UTF-8"?><Error><Code>AccessDenied</Code><Message>Access Denied</Message></Error>');

    const resp = await fetchS3('test-bucket', 'live/index.json', console);
    await scope.done();

    assert.strictEqual(resp.status, 404);
  });

  it('handles 500 from s3', async () => {
    const scope = nock('https://test-bucket.s3.fake.amazonaws.com')
      .get('/live/index.json?x-id=GetObject')
      .reply(500);

    const resp = await fetchS3('test-bucket', 'live/index.json', console);
    await scope.done();

    assert.strictEqual(resp.status, 502);
  });
});


src/index.js
/*
 * Copyright 2019 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const wrap = require('@adobe/helix-shared-wrap');
const { logger } = require('@adobe/helix-universal-logger');
const { wrap: statusWrap } = require('@adobe/helix-status');
const { Response } = require('@adobe/helix-universal');
const { cleanupHeaderValue } = require('@adobe/helix-shared-utils');
const jsonFilter = require('./json-filter.js');
const fetchS3 = require('./fetch-s3.js');

function error(log, msg, status) {
  log.error(msg);
  return new Response('', {
    status,
    headers: {
      'x-error': cleanupHeaderValue(msg),
    },
  });
}

/**
 * This is the main function
 * @param {Request} request the request object (see fetch api)
 * @param {UniversalContext} context the context of the universal serverless function
 * @returns {Response} a response
 */
async function main(request, context) {
  const { log, pathInfo: { suffix = '' } } = context;
  const sp = new URL(request.url).searchParams;
  const params = Object.fromEntries(sp.entries());
  if (params.sheet) {
    params.sheet = sp.getAll('sheet');
  }
  const {
    contentBusId,
    limit,
    offset,
    sheet,
  } = params;

  // validate params and suffix
  if (!contentBusId) {
    return error(log, 'missing contentBusId', 400);
  }
  if (!suffix.endsWith('.json')) {
    return error(log, 'only json resources supported.', 400);
  }

  // validate and create filter
  if (limit === undefined && offset === undefined && sheet === undefined) {
    return error(log, 'no filter params specified. use direct access.', 400);
  }
  const filter = jsonFilter({
    limit: limit ? Number.parseInt(limit, 10) : undefined,
    offset: offset ? Number.parseInt(offset, 10) : undefined,
    sheet,
  }, log);

  // fetch data from content bus
  const key = `${contentBusId}${suffix}`;
  const dataResponse = await fetchS3('helix-content-bus', key, log);
  if (dataResponse.status !== 200) {
    return dataResponse;
  }

  // filter and return
  return filter(dataResponse);
}

module.exports.main = wrap(main)
  .with(statusWrap)
  .with(logger);


src/fetch-s3.js
/*
 * Copyright 2021 Adobe. All rights reserved.
 * This file is licensed to you under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License. You may obtain a copy
 * of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
 * OF ANY KIND, either express or implied. See the License for the specific language
 * governing permissions and limitations under the License.
 */
const { promisify } = require('util');
const zlib = require('zlib');
const { Response } = require('@adobe/helix-universal');
const {
  S3Client,
  GetObjectCommand,
} = require('@aws-sdk/client-s3');

const gunzip = promisify(zlib.gunzip);

/**
 * Fetches content from s3
 * @param {string} bucketId bucket id
 * @param {string} key the resource key
 * @param {Logger} log the logger
 * @return {Promise<Response>>} response buffer
 */
async function fetchS3(bucketId, key, log) {
  if (!bucketId) {
    throw new Error('Unknown bucketId, cannot fetch content');
  }
  if (!key) {
    throw new Error('Unknown key, cannot fetch content');
  }

  try {
    const s3 = new S3Client();
    const res = await s3.send(new GetObjectCommand({
      Bucket: bucketId,
      Key: key,
      // todo: add timeout
    }));
    let buffer = await new Response(res.Body, {}).buffer();
    if (res.ContentEncoding === 'gzip') {
      buffer = await gunzip(buffer);
    }
    const headers = {};
    if (res.LastModified) {
      headers['last-modified'] = res.LastModified.toUTCString();
    }
    if (res.ContentType) {
      headers['content-type'] = res.ContentType;
    }
    return new Response(buffer, {
      status: 200,
      headers,
    });
  } catch (e) {
    const code = e.Code || e.$metadata.httpStatusCode;
    if (code === 'AccessDenied' || code === 404) {
      log.info(`Could not find file at ${bucketId}/${key}: ${code}`);
      return new Response('', {
        status: 404,
      });
    } else {
      log.error(`Error while fetching file from ${bucketId}/${key}: ${code}`);
      return new Response('', {
        status: 502,
        headers: {
          'x-error': `error while fetching: ${code}`,
        },
      });
    }
  }
}

module.exports = fetchS3;
